funnel/templates/event.html

```diff
@@ -46,37 +46,251 @@
+              <i class="fa fa-close loader js-abort"></i>
+            {{/checked_in}}
             </li>
             </ul>
           </td>
         </tr>
-      {%- endfor %}
+      {{/each}}
+      </script>
+      {% endraw %}
     </tbody>
   </table>
 {% endblock %}
 {% block footerscripts %}
 <script>
+  if (Modernizr.localstorage) {
+    window.Store = {};
+
+    //Reads from LocalStorage (key is 'checkin-queue' or 'cancelcheckin-queue')
+    window.Store.readFromLocalstorage = function(key) {
+      return JSON.parse(localStorage.getItem(key));
+    }
+    //Adds to LocalStorage (key is 'checkin-queue' or 'cancelcheckin-queue')
+    window.Store.addToLocalstorage = function(key, value) {
+      return localStorage.setItem(key, JSON.stringify(value))
+    }
+
+    window.Queue = {};
+
+    //Adds a participant_id to queue
+    window.Queue.enqueue = function(queueName, participant_id) {
+      var participantList = window.Store.readFromLocalstorage(queueName);
+      if(participantList) {
+        participantList.push(participant_id);
+      }
+      else {
+        participantList = [];
+        participantList.push(participant_id);
+      }
+      return window.Store.addToLocalstorage(queueName, participantList)
+    }
+
+    //Reads first item from queue
+    //Returns undefined when queue is empty or not defined
+    window.Queue.peek = function(queueName) {
+      var participantList = window.Store.readFromLocalstorage(queueName);
+      if(participantList && participantList.length) {
+        return participantList.shift();
+      }
+    }
+
+    //Reads and returns all items from queue
+    //Returns undefined when queue is empty or not defined
+    window.Queue.peekAllItems = function(queueName) {
+      var participantList = window.Store.readFromLocalstorage(queueName);
+      if(participantList && participantList.length) {
+        return participantList;
+      }
+    }
+
+    //Removes item from queue and returns true
+    //Returns undefined when item not present in queue
+    window.Queue.dequeue = function(queueName, participant_id) {
+      var participantList = window.Store.readFromLocalstorage(queueName);
+      var index = participantList ? participantList.indexOf(participant_id) : -1;
+      if(index !== -1) {
+        //Remove item from queue and add updated queue to localStorage
+        participantList.splice(index, 1);
+        window.Store.addToLocalstorage(queueName, participantList);
+        return true;
+      }
+    }
+  }
+
   $(document).ready(function() {
+    // To disable debug mode
+    Ractive.DEBUG = false;
+
+    var participantTableContent = new Ractive({
+      el: '#participants-table-content',
+      template: '#participant-row',
+      data: {
+        participants: ''
+      }
+    });
+
+    //updateQueue: If participant in "checkin-queue" has already been checked-in then it is removed from checkin queue
+    function updateQueue(queueName) {
```

Comment:
I agree, DOM is secondary source of information. But in this case, these DOM values are updated only receiving values from server. 
Any future behavior js shouldn't be allowed to alter these DOM values unless data is received from the server. The whole check-in functionality works on information from DOM right from the adding to check-in queue or cancel check-in queue.
$('#participants-table-content').on('click', '.js-checkin', function(event) {
        event.preventDefault();
        window.Queue.enqueue("checkin-queue", $(this).attr('data-pid'));
        .....